### PR TITLE
feat: add token auth mode. Fixes #9090

### LINKED ISSUE #9090: Static token authentication mode
# Summary
This would implement a new authentication method using tokens from a specific secret.

# Use Cases
This could be useful for automatically managing workflows from a different cluster or namespace.
Right now,
+ `--auth-mode server` will allow to bypass any other authentication method by design.
+ `--auth-mode client` requires access to a ServiceAccount-s secret token, which can be complicated from another cluster or even namespace.
+ `--auth-mode sso` will require a managed application and a flow to generate a token in order to authenticate.

---

**Message from the maintainers**:

Love this enhancement proposal? Give it a 👍. We prioritise the proposals with the most 👍.

### PR DESCRIPTION
**Pull request by LoricAndre**
Signed-off-by: Loric ANDRE <loric.a@padoa-group.com>
Fixes #9090

This implements a new server authentication mode, using a kube secret for static Bearer token authentication.
We require this change for cross-cluster workflow submit from other workflows without server auth mode. See #9090 for details
This is linked to [this chart update PR](https://github.com/argoproj/argo-helm/pull/1352)

### CODE REVIEW

**Comment by alexec** [comment]
Any new auth method impact security and therefore has a higher bar to acceptance. A quick visual inspection of the code tells me that I don't think it will work when your run `argo server --auth-mode=sso --auth-mode=token`.

Before we review this PR, I did not see any enhancement proposal. Can I please ask you to create an issue and explain the use base and get a few 👍 to show that people want this?

**Comment by LoricAndre** [comment]
You're right about the SSO+Token, the previous code made them incompatible. It is fixed now, and I added a test to verify it.

File: config/controller.go [hunk 1]
```diff
@@ -47,7 +47,8 @@ func parseConfigMap(cm *apiv1.ConfigMap, config *Config) error {
 			}
 		}
 	}
-	err := yaml.UnmarshalStrict([]byte(rawConfig), config)
+	// err := yaml.UnmarshalStrict([]byte(rawConfig), config)
```
Comment (alexec) [inline config/controller.go:50]: revert

File: config/config.go [hunk 2]
```diff
@@ -105,6 +105,9 @@ type Config struct {
 
 	// SSO in settings for single-sign on
 	SSO SSOConfig `json:"sso,omitempty"`
+
+	// Secret name for token auth mode
+	TokenSecretName string `json:"tokenSecretName,omitempty"`
```
Comment (alexec) [inline config/config.go:110]: I think this should be

```yaml
token:
  secretName: abc
```

The secret name should have a default value, e.g. `argo-auth-tokens`

File: config/controller_test.go [hunk 3]
```diff
@@ -30,9 +30,9 @@ func Test_parseConfigMap(t *testing.T) {
 			assert.NotEmpty(t, c.ArtifactRepository)
 		}
 	})
-	t.Run("Garbage", func(t *testing.T) {
-		c := &Config{}
-		err := parseConfigMap(&apiv1.ConfigMap{Data: map[string]string{"garbage": "garbage"}}, c)
-		assert.Error(t, err)
-	})
+	// t.Run("Garbage", func(t *testing.T) {
```
Comment (alexec) [inline config/controller_test.go:33]: revert

File: config/controller.go (hunk 1 above)
Comment (LoricAndre) [reply config/controller.go:50]: Yes, this is on the Todo before merging, this was set for our testing.

**Comment by alexec** [comment]
I think this is a good idea. I do think that we need the following:

1. This should be self-serve, I.e. look in the namespace of the request for the secret. This will allow teams in multi-tentant mode to self-serve.
2. Docs.

**Comment by LoricAndre** [comment]
@alexec about the self-serve part, implementing this would then require secrets in every workflow's namespace, instead of just the server's. This seems pretty heavy to me, I can implement a toggle for this but I'm not sure this is what you mean.

**Comment by alexec** [comment]
What we do for similar self-serve:

* Is there secret in user namespace? If so, use that.
* Otherwise, is there secret in install namespace.

Makes sense?

**Comment by LoricAndre** [comment]
I'm not sure I get what you mean by 'user namespace', since there is no user with this auth method. I can use the request's namespace instead, but this doesn't seem very useful in itself.

**Comment by alexec** [comment]
You're correct. I mean "request namespace", not "user namespace".

**Comment by alexec** [comment]
This PR is still a great idea as it makes automation easier and more secure.

**Pull request closed without merging**